src/main/docs/man/man3/DragonTalon.3
.TH "DragonTalon" 3 "Thu Oct 31 2019" "2020 Template Project" \" -*- nroff -*-
.ad l
.nh
.SH NAME
DragonTalon
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <DragonTalon\&.h>\fP
.PP
Inherits \fBIDragonMotorController\fP\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBTALON_SENSOR_TYPE\fP { \fBNO_SENSOR\fP = -1, \fBQUAD_ENCODER\fP, \fBUNKNOWN_SENSOR\fP, \fBANALOG_POT\fP, \fBANALOG_ENCODER\fP, \fBENCODER_RISING\fP, \fBENCODER_FALLING\fP, \fBCTRE_MAG_ENCODER_RELATIVE\fP, \fBCTRE_MAG_ENCODER_ABSOLUTE\fP, \fBPULSE_WIDTH\fP, \fBMAX_SENSOR_TYPES\fP }"
.br
.ti -1c
.RI "enum \fBTALON_CONTROL_MODE\fP { \fBPERCENT\fP, \fBPOSITION\fP, \fBVELOCITY\fP, \fBFOLLOWER\fP, \fBMOTION_MAGIC\fP, \fBMAX_TALON_MODES\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBDragonTalon\fP ()=delete"
.br
.ti -1c
.RI "\fBDragonTalon\fP (\fBMOTOR_CONTROLLER_USAGE\fP deviceType, int deviceID, int countsPerRev, double gearRatio)"
.br
.ti -1c
.RI "virtual \fB~DragonTalon\fP ()=default"
.br
.ti -1c
.RI "double \fBGetRotations\fP () const override"
.br
.ti -1c
.RI "double \fBGetRPS\fP () const override"
.br
.ti -1c
.RI "\fBMOTOR_CONTROLLER_USAGE\fP \fBGetType\fP () const override"
.br
.ti -1c
.RI "int \fBGetID\fP () const override"
.br
.ti -1c
.RI "void \fBSetControlMode\fP (\fBIDragonMotorController::DRAGON_CONTROL_MODE\fP mode) override"
.br
.ti -1c
.RI "void \fBSet\fP (double value) override"
.br
.ti -1c
.RI "void \fBSetRotationOffset\fP (double rotations) override"
.br
.ti -1c
.RI "void \fBSetVoltageRamping\fP (double ramping, double rampingClosedLoop=\-1) override"
.br
.ti -1c
.RI "void \fBEnableCurrentLimiting\fP (bool enabled) override"
.br
.ti -1c
.RI "void \fBEnableBrakeMode\fP (bool enabled) override"
.br
.ti -1c
.RI "void \fBSetPIDF\fP (double p, double i, double d, double f, int slot=0) override"
.br
.ti -1c
.RI "void \fBInvert\fP (bool inverted) override"
.br
.ti -1c
.RI "void \fBSetControlMode\fP (\fBTALON_CONTROL_MODE\fP mode)"
.br
.ti -1c
.RI "void \fBSetSensorInverted\fP (bool inverted)"
.br
.ti -1c
.RI "void \fBSelectClosedLoopProfile\fP (int slot, int pidIndex)"
.br
.ti -1c
.RI "int \fBConfigSelectedFeedbackSensor\fP (FeedbackDevice feedbackDevice, int pidIdx, int timeoutMs)"
.br
.ti -1c
.RI "int \fBConfigSelectedFeedbackSensor\fP (RemoteFeedbackDevice feedbackDevice, int pidIdx, int timeoutMs)"
.br
.ti -1c
.RI "int \fBConfigPeakCurrentLimit\fP (int amps, int timeoutMs)"
.br
.ti -1c
.RI "int \fBConfigPeakCurrentDuration\fP (int milliseconds, int timeoutMs)"
.br
.ti -1c
.RI "int \fBConfigContinuousCurrentLimit\fP (int amps, int timeoutMs)"
.br
.ti -1c
.RI "void \fBSetAsSlave\fP (int masterCANID)"
.br
.ti -1c
.RI "void \fBConfigMotionAcceleration\fP (float maxMotionAcceleration)"
.br
.ti -1c
.RI "void \fBConfigMotionCruiseVelocity\fP (float maxMotionCruiseVelocity)"
.br
.ti -1c
.RI "void \fBConfigPeakOutput\fP (double percentOutput)"
.br
.ti -1c
.RI "void \fBConfigNominalOutput\fP (double nominalOutput)"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
Definition at line 22 of file DragonTalon\&.h\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBDragonTalon::TALON_CONTROL_MODE\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIPERCENT \fP\fP
.TP
\fB\fIPOSITION \fP\fP
.TP
\fB\fIVELOCITY \fP\fP
.TP
\fB\fIFOLLOWER \fP\fP
.TP
\fB\fIMOTION_MAGIC \fP\fP
.TP
\fB\fIMAX_TALON_MODES \fP\fP
.PP
Definition at line 41 of file DragonTalon\&.h\&.
.SS "enum \fBDragonTalon::TALON_SENSOR_TYPE\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fINO_SENSOR \fP\fP
.TP
\fB\fIQUAD_ENCODER \fP\fP
.TP
\fB\fIUNKNOWN_SENSOR \fP\fP
.TP
\fB\fIANALOG_POT \fP\fP
.TP
\fB\fIANALOG_ENCODER \fP\fP
.TP
\fB\fIENCODER_RISING \fP\fP
.TP
\fB\fIENCODER_FALLING \fP\fP
.TP
\fB\fICTRE_MAG_ENCODER_RELATIVE \fP\fP
.TP
\fB\fICTRE_MAG_ENCODER_ABSOLUTE \fP\fP
.TP
\fB\fIPULSE_WIDTH \fP\fP
.TP
\fB\fIMAX_SENSOR_TYPES \fP\fP
.PP
Definition at line 26 of file DragonTalon\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "DragonTalon::DragonTalon ()\fC [delete]\fP"

.SS "DragonTalon::DragonTalon (\fBMOTOR_CONTROLLER_USAGE\fP deviceType, int deviceID, int countsPerRev, double gearRatio)"

.PP
Definition at line 3 of file DragonTalon\&.cpp\&.
.SS "virtual DragonTalon::~DragonTalon ()\fC [virtual]\fP, \fC [default]\fP"

.SH "Member Function Documentation"
.PP 
.SS "int DragonTalon::ConfigContinuousCurrentLimit (int amps, int timeoutMs)"

.PP
Definition at line 284 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::ConfigMotionAcceleration (float maxMotionAcceleration)"

.PP
Definition at line 173 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::ConfigMotionCruiseVelocity (float maxMotionCruiseVelocity)"

.PP
Definition at line 183 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::ConfigNominalOutput (double nominalOutput)"

.PP
Definition at line 201 of file DragonTalon\&.cpp\&.
.SS "int DragonTalon::ConfigPeakCurrentDuration (int milliseconds, int timeoutMs)"

.PP
Definition at line 266 of file DragonTalon\&.cpp\&.
.SS "int DragonTalon::ConfigPeakCurrentLimit (int amps, int timeoutMs)"

.PP
Definition at line 248 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::ConfigPeakOutput (double percentOutput)"

.PP
Definition at line 192 of file DragonTalon\&.cpp\&.
.SS "int DragonTalon::ConfigSelectedFeedbackSensor (FeedbackDevice feedbackDevice, int pidIdx, int timeoutMs)"

.PP
Definition at line 210 of file DragonTalon\&.cpp\&.
.SS "int DragonTalon::ConfigSelectedFeedbackSensor (RemoteFeedbackDevice feedbackDevice, int pidIdx, int timeoutMs)"

.PP
Definition at line 229 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::EnableBrakeMode (bool enabled)\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 124 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::EnableCurrentLimiting (bool enabled)\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 119 of file DragonTalon\&.cpp\&.
.SS "int DragonTalon::GetID () const\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 152 of file DragonTalon\&.cpp\&.
.SS "double DragonTalon::GetRotations () const\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 19 of file DragonTalon\&.cpp\&.
.SS "double DragonTalon::GetRPS () const\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 24 of file DragonTalon\&.cpp\&.
.SS "\fBMOTOR_CONTROLLER_USAGE\fP DragonTalon::GetType () const\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 147 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::Invert (bool inverted)\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 137 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::SelectClosedLoopProfile (int slot, int pidIndex)"

.PP
Definition at line 163 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::Set (double value)\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 64 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::SetAsSlave (int masterCANID)"

.PP
Definition at line 302 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::SetControlMode (\fBIDragonMotorController::DRAGON_CONTROL_MODE\fP mode)\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 29 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::SetControlMode (\fBDragonTalon::TALON_CONTROL_MODE\fP mode)"

.PP
Definition at line 53 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::SetPIDF (double p, double i, double d, double f, int slot = \fC0\fP)\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 129 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::SetRotationOffset (double rotations)\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 101 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::SetSensorInverted (bool inverted)"

.PP
Definition at line 142 of file DragonTalon\&.cpp\&.
.SS "void DragonTalon::SetVoltageRamping (double ramping, double rampingClosedLoop = \fC\-1\fP)\fC [override]\fP, \fC [virtual]\fP"

.PP
Implements \fBIDragonMotorController\fP\&.
.PP
Definition at line 108 of file DragonTalon\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for 2020 Template Project from the source code\&.


src/main/cpp/hw/DragonTalon.cpp
#include <hw/DragonTalon.h>
#include <hw/usages/MotorControllerUsage.h>
#include <frc/SpeedController.h>
#include <memory>

using namespace frc;
using namespace std;
using namespace ctre::phoenix::motorcontrol::can;
using namespace ctre::phoenix::motorcontrol;
using namespace ctre::phoenix;

DragonTalon::DragonTalon
(
	MotorControllerUsage::MOTOR_CONTROLLER_USAGE deviceType, 
	int deviceID, 
	int countsPerRev, 
	double gearRatio ) : m_talon( make_shared<WPI_TalonSRX>(deviceID)),
    					 m_controlMode(TALON_CONTROL_MODE::PERCENT),
						 m_type(deviceType),
						 m_id(deviceID),
						 m_countsPerRev(countsPerRev),
    					 m_tickOffset(0),
						 m_gearRatio(gearRatio)
{
	// m_tickOffset
	// m_talon->GetSelectedSensorPo
	m_tickOffset = m_talon->GetSelectedSensorPosition();
}

double DragonTalon::GetRotations() const
{
    return ((m_talon->GetSelectedSensorPosition() - m_tickOffset) / (double) m_countsPerRev) * m_gearRatio;
}

double DragonTalon::GetRPS() const
{
    return m_talon->GetSelectedSensorVelocity() / (double) m_countsPerRev;
}

void DragonTalon::SetControlMode(IDragonMotorController::DRAGON_CONTROL_MODE mode)
{ 
    switch (mode)
    {
        case IDragonMotorController::DRAGON_CONTROL_MODE::PERCENT_OUTPUT:
            DragonTalon::SetControlMode(DragonTalon::TALON_CONTROL_MODE::PERCENT);
        break;

        case IDragonMotorController::DRAGON_CONTROL_MODE::ROTATIONS:
            DragonTalon::SetControlMode(DragonTalon::TALON_CONTROL_MODE::POSITION);
        break;
        
        case IDragonMotorController::DRAGON_CONTROL_MODE::RPS:
            DragonTalon::SetControlMode(DragonTalon::TALON_CONTROL_MODE::VELOCITY);
        break;
        
        default:
            // bad place to be
			printf("SCREAMIN!!!!!111 invalid controlmode set in DragonTalon SetControlMode\n");
            DragonTalon::SetControlMode(DragonTalon::TALON_CONTROL_MODE::PERCENT);
        break;
    }
}

shared_ptr<SpeedController> DragonTalon::GetSpeedController() const
{
	return m_talon;
}

void DragonTalon::SetControlMode(DragonTalon::TALON_CONTROL_MODE mode)
{
    if (m_controlMode != mode)
    {
        m_controlMode = mode;
		// printf("mode changed \n");
        // DragonTalon::Set(0);
    }
    
}

void DragonTalon::Set(double value)
{
    switch (m_controlMode)
    {
        case TALON_CONTROL_MODE::PERCENT:
            m_talon->Set(ctre::phoenix::motorcontrol::ControlMode::PercentOutput, value);
        break;

        case TALON_CONTROL_MODE::POSITION:
		// return ((m_talon->GetSelectedSensorPosition() - m_tickOffset) / (double) m_countsPerRev) * m_gearRatio;
            // m_talon->Set(ctre::phoenix::motorcontrol::ControlMode::Position, (value + m_tickOffset) / ((double) m_countsPerRev * m_gearRatio));
			m_talon->Set(ctre::phoenix::motorcontrol::ControlMode::Position, (value * m_countsPerRev / m_gearRatio) + m_tickOffset);
        break;

        case TALON_CONTROL_MODE::VELOCITY:
            m_talon->Set(ctre::phoenix::motorcontrol::ControlMode::Velocity, value * m_gearRatio / 600.0);
        break;

        //TODO: 
        // case TALON_CONTROL_MODE::RPS:
        //     m_talon->Set(ctre::phoenix::motorcontrol::ControlMode::Velocity, value / m_countsPerRev);    
        // break;

        case TALON_CONTROL_MODE::FOLLOWER:
            m_talon->Set(ctre::phoenix::motorcontrol::ControlMode::Follower, value);
        break;

        case TALON_CONTROL_MODE::MOTION_MAGIC:
            m_talon->Set(ctre::phoenix::motorcontrol::ControlMode::MotionMagic, (value * m_countsPerRev / m_gearRatio) + m_tickOffset);
        break;

        default:
            m_talon->Set(ctre::phoenix::motorcontrol::ControlMode::PercentOutput, 0);
        break;
    }
}

void DragonTalon::SetRotationOffset(double rotations)
{
    // m_tickOffset = (rotations * m_countsPerRev * m_gearRatio);
	double newRotations = -rotations + DragonTalon::GetRotations();
	m_tickOffset += (int) (newRotations * m_countsPerRev / m_gearRatio);
}

void DragonTalon::SetVoltageRamping(double ramping, double rampingClosedLoop)
{
    m_talon->ConfigOpenloopRamp(ramping);

	if (rampingClosedLoop >= 0)
	{
		m_talon->ConfigClosedloopRamp(rampingClosedLoop);
	}
}


void DragonTalon::EnableCurrentLimiting(bool enabled)
{
    m_talon->EnableCurrentLimit(enabled);
}

void DragonTalon::EnableBrakeMode(bool enabled)
{
    m_talon->SetNeutralMode(enabled ? ctre::phoenix::motorcontrol::NeutralMode::Brake : ctre::phoenix::motorcontrol::NeutralMode::Coast);
}

void DragonTalon::SetPIDF(double p, double i, double d, double f, int slot)
{
    m_talon->Config_kP(slot, p);
    m_talon->Config_kI(slot, i);
    m_talon->Config_kD(slot, d);
    m_talon->Config_kF(slot, f);
}

void DragonTalon::Invert(bool inverted)
{
    m_talon->SetInverted(inverted);
}

void DragonTalon::SetSensorInverted(bool inverted)
{
    m_talon->SetSensorPhase(inverted);
}

MotorControllerUsage::MOTOR_CONTROLLER_USAGE DragonTalon::GetType() const
{
	return m_type;
}

int DragonTalon::GetID() const
{
	return m_id;
}

//------------------------------------------------------------------------------
// Method:		SelectClosedLoopProfile
// Description:	Selects which profile slot to use for closed-loop control
// Returns:		void
//------------------------------------------------------------------------------
void DragonTalon::SelectClosedLoopProfile
(
	int	   slot,			// <I> - profile slot to select
	int    pidIndex			// <I> - 0 for primary closed loop, 1 for cascaded closed-loop
)
{
	m_talon->SelectProfileSlot( slot, pidIndex );
}

//hopefully this is correct
void DragonTalon::ConfigMotionAcceleration
(
	float maxMotionAcceleration
)
{
    int sensorUnitsPer100msPerSec = static_cast<int>( (maxMotionAcceleration / 360.0) * (m_countsPerRev / 10.0) / m_gearRatio );
	m_talon->ConfigMotionAcceleration( sensorUnitsPer100msPerSec, 0);
}

//need to change
void DragonTalon::ConfigMotionCruiseVelocity
(
	float maxMotionCruiseVelocity
)
{
    int sensorUnitsPer100msPerSec = static_cast<int>( (maxMotionCruiseVelocity / 360.0) * (m_countsPerRev / 10.0) / m_gearRatio );
	m_talon->ConfigMotionCruiseVelocity( sensorUnitsPer100msPerSec, 0);
}

void DragonTalon::ConfigPeakOutput
(
	double percentOutput
)
{
	m_talon->ConfigPeakOutputForward(percentOutput);
	m_talon->ConfigPeakOutputReverse(-percentOutput);
}

void DragonTalon::ConfigNominalOutput
(
	double nominalOutput
)
{
	m_talon->ConfigNominalOutputForward(nominalOutput);
	m_talon->ConfigNominalOutputReverse(-nominalOutput);
}

int DragonTalon::ConfigSelectedFeedbackSensor
(
	FeedbackDevice feedbackDevice,
	int pidIdx,
	int timeoutMs
)
{
	int error = 0;
	if ( m_talon != nullptr )
	{
		error = m_talon->ConfigSelectedFeedbackSensor( feedbackDevice, pidIdx, timeoutMs );
	}
	else
	{
//		printf( "DragonTalon::ConfigSelectedFeedbackSensor: m_talon not initialized \n" );
	}
	return error;
}

int DragonTalon::ConfigSelectedFeedbackSensor
(
	RemoteFeedbackDevice feedbackDevice,
	int pidIdx,
	int timeoutMs
)
{
	int error = 0;
	if ( m_talon != nullptr )
	{
		error = m_talon->ConfigSelectedFeedbackSensor( feedbackDevice, pidIdx, timeoutMs );
	}
	else
	{
//		printf( "DragonTalon::ConfigSelectedFeedbackSensor: m_talon not initialized \n" );
	}
	return error;
}

int DragonTalon::ConfigPeakCurrentLimit
(
	int amps,
	int timeoutMs
)
{
	int error = 0;
	if ( m_talon != nullptr )
	{
		error = m_talon->ConfigPeakCurrentLimit( amps, timeoutMs );
	}
	else
	{
//		printf( "DragonTalon::ConfigPeakCurrentLimit: m_talon not initialized \n" );
	}
	return error;
}

int DragonTalon::ConfigPeakCurrentDuration
(
	int milliseconds,
	int timeoutMs
)
{
	int error = 0;
	if ( m_talon != nullptr )
	{
		error = m_talon->ConfigPeakCurrentDuration( milliseconds, timeoutMs );
	}
	else
	{
//		printf( "DragonTalon::ConfigPeakCurrentDuration: m_talon not initialized \n" );
	}
	return error;
}

int DragonTalon::ConfigContinuousCurrentLimit
(
	int amps,
	int timeoutMs
)
{
	int error = 0;
	if ( m_talon != nullptr )
	{
		error = m_talon->ConfigContinuousCurrentLimit( amps, timeoutMs );
	}
	else
	{
//		printf( "DragonTalon::ConfigContinuousCurrentLimit: m_talon not initialized \n" );
	}
	return error;
}

void DragonTalon::SetAsSlave
(
    int         masterCANID         // <I> - master motor
)
{
    m_talon->Set( ControlMode::Follower, masterCANID );
}

